### PR TITLE
For #19005: new tab three-dot menu sync sign in 

### PR DESCRIPTION
**Pull request by eliserichards**
For #19005 

The sync item in the new tab three-dot menu should reflect the tabs tray rewrite. 
* If synced tabs are in the tabs tray, we show our new "sync sign in" item.
* Sync sign in clicks navigate to account settings

  image

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt [hunk 1]
```diff
@@ -64,6 +66,9 @@ class HomeMenu(
 
     private val shouldUseBottomToolbar = context.settings().shouldUseBottomToolbar
 
+    private var signedInToFxA = false
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:69]: `signedInToFxA` doesn't seem to be used for anything

File: app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt [hunk 2]
```diff
@@ -289,14 +294,30 @@ class HomeMenu(
...
             R.drawable.ic_synced_tabs,
             primaryTextColor
         ) {
             onItemTapped.invoke(Item.SyncTabs)
         }
 
+        val syncItemTitle =
+            if (accountManager.accountProfile()?.email != null) {
+                signedInToFxA = true
+                accountManager.accountProfile()?.email!!
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:308]: You can avoid the !! by setting `accountManager.accountProfile()?.email` to a variable
```suggestion
        val email = accountManager.accountProfile()?.email
        val syncItemTitle = email ?: context.getString(R.string.sync_menu_sign_in)
```

File: app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt [hunk 3]
```diff
@@ -289,14 +294,30 @@ class HomeMenu(
...
+        val syncedTabsItem = BrowserMenuImageText(
+            context.getString(R.string.synced_tabs),
             R.drawable.ic_synced_tabs,
             primaryTextColor
         ) {
             onItemTapped.invoke(Item.SyncTabs)
         }
 
+        val syncItemTitle =
+            if (accountManager.accountProfile()?.email != null) {
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:306]: I think we want to check with someone understands the FxA code a bit more to check if this is good enough because there are at least 3 states iirc. (Signed in, Need to reconnect, Sign in to sync). If we don't check if the account is authenticated, I think we can run into the 2nd state where the user actually needs to reconnect.

**Comment by eliserichards** [comment]
These are all fixed in the other PR ^

**Comment by eliserichards** [comment]
I'll rebase this one once #19035 lands

**Comment by gabrielluong** [comment]
https://console.firebase.google.com/project/moz-fenix/testlab/histories/bh.66b7091e15d53d45/matrices/8581703720764238539/executions/bs.48f00f839178dc7c/testcases/2/test-cases

> java.lang.AssertionError
at org.junit.Assert.fail(Assert.java:87)
at org.junit.Assert.assertTrue(Assert.java:42)
at org.junit.Assert.assertNotNull(Assert.java:713)
at org.junit.Assert.assertNotNull(Assert.java:723)
at org.mozilla.fenix.helpers.ext.WaitNotNullKt.waitNotNull(WaitNotNull.kt:20)
at org.mozilla.fenix.ui.robots.ThreeDotMenuMainRobot$Transition.openSyncedTabs(ThreeDotMenuMainRobot.kt:161)
at org.mozilla.fenix.ui.SmokeTest.openMainMenuSyncedTabsItemTest(SmokeTest.kt:233)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
at androidx.test.internal.runner.junit4.statement.RunAfters.evaluate(RunAfters.java:61)
at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:134)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:395)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2145)

**Review by gabrielluong — DISMISSED** [review]
Let's fix the test before review.

File: app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt [hunk 4]
```diff
@@ -236,6 +226,17 @@ class HomeMenu(
...
+        val authenticatedAccount = accountManager.authenticatedAccount() != null
+        val email = accountManager.accountProfile()?.email
+
+        return if (authenticatedAccount && email != null) {
+            email
+        } else {
+            context.getString(R.string.sync_menu_sign_in)
+        }
+    }
+
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:239]: Was this move intentional?

Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:239]: Yes

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (gabrielluong) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt:1]: Not sure if i am seeing this correctly, but I think there's an extra space.

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/ThreeDotMenuMainRobot.kt [hunk 6]
```diff
@@ -393,221 +403,226 @@ class ThreeDotMenuMainRobot {
         }
     }
 }
-private fun threeDotMenuRecyclerView() =
-    onView(withId(R.id.mozac_browser_menu_recyclerView))
 
-private fun threeDotMenuRecyclerViewExists() {
-    threeDotMenuRecyclerView().check(matches(isDisplayed()))
-}
+    private fun threeDotMenuRecyclerView() =
```
Comment (gabrielluong) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/ThreeDotMenuMainRobot.kt:407]: How come we indented all these functions below this? These functions aren't inside of the ThreeDotMenuMainRobot class so, I don't see any particular reason this should be indented. I did go to this file and do reformat code in Android Studio to see if this was a lint issue. I imagine we want to revert this change since the change is unnecessary.

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        }
+    }
+}
+
+private fun goBackButton() =
+    onView(allOf(withContentDescription("Navigate up")))
+
+private fun assertAccountSettingsMenuHeader() {
+    // Replaced with the new string here, the test is assuming we are NOT signed in
+    // Sync tests in SettingsSyncTest are still TO-DO, so I'm not sure that we have a test for signing into Sync
```
Comment (gabrielluong) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt:43]: I don't think we necessarily want this personal monologue in the codebase. 😜

I would suggest rewording to just indicate the remaining work or just remove it completely and have an issue filed.

**Review by gabrielluong — CHANGES_REQUESTED** [review]

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt (hunk 7 above)
Comment (eliserichards) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt:43]: This is not my comment. It is duping the comment from here https://github.com/mozilla-mobile/fenix/blob/48460ce42cc57d033b3ad1746ad6ec63339fa393/app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncedTabsRobot.kt#L42 because it is relevant to this test too. The UI test peeps are handling SettingsSyncTest

Comment (eliserichards) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt:43]: I should have commented that initially, sorry!!

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt (hunk 5 above)
Comment (eliserichards) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt:1]: I can't see one

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/ThreeDotMenuMainRobot.kt (hunk 6 above)
Comment (eliserichards) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/ThreeDotMenuMainRobot.kt:407]: it was a lint fix. I can change it back

**Comment by eliserichards** [comment]
Rebased with latest from master

**Review by gabrielluong — APPROVED** [review]
Nice, also appreciate the fact that we added the test for this.

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/ThreeDotMenuMainRobot.kt (hunk 6 above)
Comment (gabrielluong) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/ThreeDotMenuMainRobot.kt]: Weird, my reformat didn't indent everything. It did add a new line here tho. Looking at it visually, I still wouldn't expect the indent since these functions are not in a scope/block.

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt (hunk 7 above)
Comment (gabrielluong) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/SyncSignInRobot.kt:43]: I think we should consider removing the comment if we aren't planning any additional actions here or have any issue filed on what we want to achieve and do here. I think this goes back to the code ownership theme where we do want to understand the code we are adding even if it's copy-pasta, and just be thoughtful about it.
